prometheus/promhttp/instrument_server.go

```diff
@@ -0,0 +1,363 @@
+// will also contain tooling to instrument instances of http.Handler and
+// http.RoundTripper.
+//
+// promhttp.Handler acts on the prometheus.DefaultGatherer. With HandlerFor,
+// you can create a handler for a custom registry or anything that implements
+// the Gatherer interface. It also allows to create handlers that act
+// differently on errors or allow to log errors.
+package promhttp
+
+import (
+	"bufio"
+	"io"
+	"net"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+func InFlight(g prometheus.Gauge, next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		g.Inc()
+		next.ServeHTTP(w, r)
+		g.Dec()
+	})
+}
+
+// Latency accepts an ObserverVec interface and an http.Handler, returning a
+// new http.Handler that wraps the supplied http.Handler. The provided
+// ObserverVec must be registered in a registry in order to be used.
+// If the wrapped http.Handler has not set a status code, i.e. the value is
+// currently 0, the supplied ObserverVec will report a 200.
+// The instance labels "code" and are supported on the provided ObserverVec.
+func Latency(obs prometheus.ObserverVec, next http.Handler) http.Handler {
+	// Maybe include path? Tell people to use a const label for now.
+	var code bool
+	if _, err := obs.GetMetricWith(prometheus.Labels{"code": "0"}); err == nil {
+		code = true
+	}
+	// TODO: How to delete these labels?
+	// c.Delete(labels)
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		now := time.Now()
+		delegate := &responseWriterDelegator{ResponseWriter: w}
+		var label string
+
+		next.ServeHTTP(delegate, r)
+
+		if code {
+			label = sanitizeCode(delegate.status)
+		}
+		obs.WithLabelValues(label).Observe(time.Since(now).Seconds())
+	})
+}
+
+// Latency accepts an ObserverVec interface and an http.Handler, returning a
+// new http.Handler that wraps the supplied http.Handler. The provided
+// CounterVec must be registered in a registry in order to be used.
+// If the wrapped http.Handler has not set a status code, i.e. the value is
+// currently 0, the supplied counter will report a 200.
+// The instance labels "code" and "method" are supported on the provided
+// CounterVec.
+func Counter(counter *prometheus.CounterVec, next http.Handler) http.Handler {
+	var (
+		code   bool
+		method bool
+		err    error
+	)
+
+	// TODO(beorn7): Remove this hacky way to check for instance labels
+	// once Descriptors can have their dimensionality queried.
+	if _, err = counter.GetMetricWith(prometheus.Labels{"code": "0", "method": "get"}); err == nil {
+		code = true
+		method = true
+	} else if _, err = counter.GetMetricWith(prometheus.Labels{"method": "get"}); err == nil {
+		method = true
+	} else if _, err = counter.GetMetricWith(prometheus.Labels{"code": "0"}); err == nil {
+		code = true
+	} else if _, err = counter.GetMetricWith(prometheus.Labels{}); err != nil {
```

Comment:
The current approach has the problem to add a metric to the …Vec that might never be used. That's visible for the user and thus confusing (and arguably wrong). My suggestion above doesn't touch the …Vec, so the consequences of the hack don't leak. It will also be at one place in the code, so once there are better ways to check a `Desc`, we can simply change it there.